packages/anoncreds/src/anoncreds-rs/AnonCreds2023DataIntegrityService.ts

```diff
@@ -0,0 +1,580 @@
+    const predicates: {
+      predicateType: AnonCredsRequestedPredicate['p_type']
+      predicateValue: AnonCredsRequestedPredicate['p_value']
+    }[] = []
+
+    const supportedJsonSchemaNumericRangeProperties: Record<string, AnonCredsRequestedPredicate['p_type']> = {
+      exclusiveMinimum: '>',
+      exclusiveMaximum: '<',
+      minimum: '>=',
+      maximum: '<=',
+    }
+
+    for (const [key, value] of Object.entries(predicateFilter)) {
+      if (key === 'type') continue
+
+      const predicateType = supportedJsonSchemaNumericRangeProperties[key]
+      if (!predicateType) throw new CredoError(`Unsupported predicate filter property '${key}'`)
+      predicates.push({
+        predicateType,
+        predicateValue: value,
+      })
+    }
+
+    return predicates
+  }
+
+  private getRevocationMetadataForCredentials = async (
+    agentContext: AgentContext,
+    credentialsWithMetadata: CredentialWithMetadata[]
+  ) => {
+    const revocationMetadataFetchPromises = credentialsWithMetadata
+      .filter((cwm) => cwm.nonRevoked)
+      .map(async (credentialWithMetadata) => {
+        const { revocationRegistryIndex, revocationRegistryId, timestamp } = AnonCredsW3cCredential.fromJson(
+          credentialWithMetadata.credential
+        )
+        return await this.getRevocationMetadata(agentContext, {
+          nonRevokedInterval: credentialWithMetadata.nonRevoked as AnonCredsNonRevokedInterval,
+          timestamp: timestamp,
+          revocationRegistryId,
+          revocationRegistryIndex,
+        })
+      })
+
+    return await Promise.all(revocationMetadataFetchPromises)
+  }
+
+  private getClaimNameForField(field: FieldV2) {
+    if (!field.path) throw new CredoError('Field path is required')
+    // fixme: could the path start otherwise?
+    const baseClaimPath = '$.credentialSubject.'
+    const claimPaths = field.path.filter((path) => path.startsWith(baseClaimPath))
+    if (claimPaths.length === 0) return undefined
+
+    // FIXME: we should iterate over all attributes of the schema here and check if the path is valid
+    // see https://identity.foundation/presentation-exchange/#presentation-definition
+    const claimNames = claimPaths.map((path) => path.slice(baseClaimPath.length))
+    const propertyName = claimNames[0]
+
+    return propertyName
+  }
+
+  public createAnonCredsProofRequestAndMetadata = async (
+    agentContext: AgentContext,
+    presentationDefinition: PresentationDefinitionV1 | PresentationDefinitionV2,
+    presentationSubmission: PresentationSubmission,
+    credentials: JsonObject[]
+  ) => {
+    const credentialsProve: CredentialProve[] = []
+    const schemaIds = new Set<string>()
+    const credentialDefinitionIds = new Set<string>()
+    const credentialsWithMetadata: CredentialWithMetadata[] = []
+
+    const hash = Hasher.hash(TypedArrayEncoder.fromString(presentationDefinition.id), 'sha-256')
+    const nonce = new BigNumber(hash).toString().slice(0, 32)
```

Comment:
BTW -- I think this should be documented as it can lead to issues with verification if the nonce is different during verification.

I think we should use the `challenge` the presentation request: https://github.com/hyperledger/aries-rfcs/blob/main/features/0510-dif-pres-exch-attach/README.md#request-presentation-attachment-format 

We already pass this to the `w3cCredentialService.verifyPresentation` in the `DifPresentationExchangeProofFormatService` and I think we can pass it here as well.

MAYBE we need to define that if the nonce is not a 32 bytes long number, you need to hash the `challenge` from DIF PEX, and take the first 32 bytes of the bignumber of the hash (I don't know exactly what bignumber does).

Opened an issue in AnonCreds spec as well: https://github.com/hyperledger/anoncreds-spec/issues/198

Reply:
I think this is great 👍